### ISSUE TITLE
HID Windows Truncates Data

### ISSUE THREAD

**Issue by tresf** [qzind/tray#286]
With short data streams, everything seems to be OK, but if you scan (for example) a large QR code, the data gets corrupt on Windows using PureJavaHIDAPI:

```js
// HID4Java - Windows
"02", "38", "00", "00", "00", "2c", "20", "71", "75", "69", "73", "20", "6e", "6f", "73", "74", "72", "75", "64", "20", "65", "78", "65", "72", "63", "69", "74", "61", "74", "69", "6f", "6e", "20", "75", "6c", "6c", "61", "6d", "63", "6f", "20", "6c", "61", "62", "6f", "72", "69", "73", "20", "6e", "69", "73", "69", "20", "75", "74", "20", "61", "6c", "69", "..."

// HID4Java - Linux
"02", "36", "5d", "51", "31", "75", "6c", "70", "61", "20", "71", "75", "69", "20", "6f", "66", "66", "69", "63", "69", "61", "20", "64", "65", "73", "65", "72", "75", "6e", "74", "20", "6d", "6f", "6c", "6c", "69", "74", "20", "61", "6e", "69", "6d", "20", "69", "64", "20", "65", "73", "74", "20", "6c", "61", "62", "6f", "72", "75", "6d", "2e", "0d", "00", "00"

// PureJavaHIDAPI - Windows
"00", "13", "00", "00", "00", "69", "6d", "20", "69", "64", "20", "65", "73", "74", "20", "6c", "61", "62", "6f", "72", "75", "6d", "2e", "0d", "00", "00", "00", "00", "00", "00", "00", "00", "00", "00", "00", "00", "00", "00", "00", "00", "00", "00", "00", "00", "00", "00", "00", "00", "00", "00", "00", "00", "00", "00", "00", "00", "00", "00", "00", "00", "..."
```

![image](https://user-images.githubusercontent.com/6345473/37636336-f60e4f3c-2bd7-11e8-811d-2239b4f1e22a.png)

**Comment by tresf** [comment]
Ok... so it appears we just send the latest feature report if the data is streaming faster than the read interval... 

Putting a debug on `PJHA_HidIO::onInputReport`, I get the following data from the above QR code:

```
38000000657475722061646970697363696E6720656C69742C2073656420646F20656975736D6F642074656D706F7220696E6369646964756E74207500000100
3800000074206C61626F726520657420646F6C6F7265206D61676E6120616C697175612E20557420656E696D206164206D696E696D2076656E69616D00000100
380000002C2071756973206E6F737472756420657865726369746174696F6E20756C6C616D636F206C61626F726973206E69736920757420616C697100000100
3800000075697020657820656120636F6D6D6F646F20636F6E7365717561742E2044756973206175746520697275726520646F6C6F7220696E20726500000100
3800000070726568656E646572697420696E20766F6C7570746174652076656C697420657373652063696C6C756D20646F6C6F72652065752066756700000100
38000000696174206E756C6C612070617269617475722E204578636570746575722073696E74206F6363616563617420637570696461746174206E6F00000100
380000006E2070726F6964656E742C2073756E7420696E2063756C706120717569206F666669636961206465736572756E74206D6F6C6C697420616E00000100
13000000696D20696420657374206C61626F72756D2E0D0000000000000000000000000000000000000000000000000000000000000000000000000000000000
```

Whereas QZ Tray only reports the following (only the last line):

```
13000000696d20696420657374206c61626f72756d2e0d000000000000000000000000000000000000000000000000000000000000000000000000
```

**Comment by bberenz** [comment]
I've added a basic buffer queue to PJHA classes under 024a908581151e015ea57ba6a4857cafef9c1db6. Not sure how this impacts scales (which constantly stream their data) yet though.

**Comment by tresf** [comment]
Thanks @bberenz however something's still not right.  Do you need to do a `System.arraycopy(...)` so that the underlying buffer doesn't clear itself?

**Comment by tresf** [comment]
`System.arraycopy(...)` did the trick.  To simulate the auto-flushing buffer behavior of HID4Java, we'll also be adding a 32-chunk array buffer for PureJavaHIDAPI's streaming mode.

**Comment by bberenz** [comment]
Array copy and buffer size implemented under c4134f49ee1f26b870568a4200da8e7c2ed83daf
JS api rewrite to allow additional selectors (#288) is included in this branch as well.
